### ISSUE TITLE
Long time between execution checks triggering TTL Warnings

### ISSUE THREAD

**Issue by andybCDL** [sensu/sensu#1552]
Hi folks

Apologies for posting on github but I couldn't find much of a forum for this, we have sensu configured in our environments via puppet, we then have a simple python script which rolls up the events via the API into a slack alert.

We are getting constantly spammed by warnings for checks which have failed the TTL, example;

DEV - WARNING: check_disks on HOSTNAME
----------------
Last check execution was 1000 seconds ago

In the above example, the interval and TTL are set as follows;

  # Check HDD Usage
  sensu::check { 'check_disks':
    command     => 'PATH=$PATH:/usr/lib64/nagios/plugins check_disk -w 100 -c 50 -W 8% -K 3% -m -e -n -X tmpfs -X devtmpfs',
    interval    => 600,
    ttl         => 900,
    subscribers => 'base',
    occurrences => 2,
    handle      => true,
    handlers    => ['elasticsearch'],

when i run these checks manually on the box, they return a response immediately. So my question, what would cause these checks to take longer than the TTL? Any suggestions? Aside from increasing the TTL in relation to the interval.

Thanks
Andy

**Comment by portertech** [comment]
Hi @andybCDL, do the Sensu client logs show when executions are being attempted? They should provide an idea of when they are being executed and for how long. I wonder if the check is getting hung up on a device etc.

**Comment by kobbe** [comment]
I also have this problem sometimes, even on checks that run on the server. I will investigate the logs for one of these occurrences and report back in a few days!

**Comment by agoddard** [comment]
hi @andybCDL & @kobbe you should have lines in your client logs showing execution time ("duration"), can you see if this value is what you're expecting, or whether the check is indeed timing out?

**Comment by gwagner** [comment]
@andybCDL we see this kind of thing in our environment when the result processing is being overwhelmed by things like, slow handlers for metric processing, slow handlers for alerting.  Handlers spin up a BG thread for every single thing they do.

We also see this when the sensu server itself is overwhelmed.  TTL validation is a BG process on the same server that runs scheduling and theoretically result processing.  We actually turned TTL checks off for the time being while the Sensu team (btw keep up the good work sensu team) adds the ability to split this kind of thing off from the leader.

So in short, generally the check itself is not the problem.  The problem for us has always been in the way sensu server processing results and manages its background threads.

**Comment by agoddard** [comment]
@andybCDL any luck with the `duration` data? Do you see your servers until high load as well?

**Comment by portertech** [comment]
Closing due to inactivity.

**Comment by kobbe** [comment]
I got the problem a few days ago again and this time I looked into it. 

Sensu 0.26.5 running on linux. It is a server and client running on the same machine.

My server logs:
```
...
19:10:48    door-status     publishing check request                                        
19:11:36                    determining stale clients                                       
19:11:36                    determining stale check results                                 
19:11:48    door-status     publishing check request                                        
19:15:56                    determining stale clients                                       
19:15:56                    determining stale check results                                 
19:15:56    keepalive       processing event "No keepalive sent from client for 242 seconds"
19:15:56    door-status     process event Last check execution was 248 seconds ago          
19:15:56    door-status     handler output Last check execution was 248 seconds ago         
19:16:08    door-status     publishing check request                                        
19:16:09    door-status     processing event "OK"  
...    
```                                     


My client logs:
```
...
door-status check request         19:10:48
door-status publish result: "OK"  19:10:49
door-status check request         19:11:18
door-status publish result: "OK"  19:11:19
door-status check request         19:11:48
door-status publish result: "OK"  19:11:49
door-status check request         19:16:08
door-status publish result: "OK"  19:16:09
...
```

The sensu api have loggs for the times:
```
...
2017-03-22T19:11:57.400815
2017-03-22T19:11:57.459604
2017-03-22T19:11:57.507502
2017-03-22T19:11:57.563093
2017-03-22T19:11:57.603896
2017-03-22T19:11:57.645973
2017-03-22T19:15:52.013169
2017-03-22T19:15:52.070191
2017-03-22T19:15:52.124744
2017-03-22T19:15:52.181029
2017-03-22T19:15:52.241401
...
```


19.16 I got an email: door-status: Last check execution was 248 seconds ago"

Lastly, my journalctl shows other stuff happening during the sensu outage so the server was not down during that time or anything like that.

I seems like sensu did nothing for almost 4 minutes. Despite having check that should be schedules every 30s. Any ideas of more logs that I need to check? Like rabbitmq or whatnot?
